### DISCUSSION TITLE
Data binding - nested array  - strings vs integer casting

### DISCUSSION THREAD

**Discussion by levi730** [livewire/livewire#3595] [7. Bugs (archived)]
Hello!

I've been struggling with an issue, and just figured out a workaround.  I'm not sure if this is some other known issue or not, as there is lots of conversation about the key words involved.

I'm working on a component that displays nested array data in a loop.  This data has checkboxes for selection and further processing. I want the component to load with all checkboxes checked, and for the user to uncheck boxes if they don't want to affect the record.  My code works in the opposite situation (nothing is checked at mount).  

Here's a Laravel Playground snippet showing the issue in it's simplest form:
https://laravelplayground.com/#/snippets/8910c7c3-de0e-4015-b6c1-c2656e08165a

Here is the relevant code in case the Playground goes away:

Livewire Component:
```
<?php

use Livewire\Component;

class HelloWorld extends Component
{
    public $selections = [];
    public $data = [
      [ 'id' => 1, 'text' => 'Foo'],
      [ 'id' => 2, 'text' => 'Bar'],
      [ 'id' => 3, 'text' => 'Baz'],
      [ 'id' => 4, 'text' => 'Quux'],
    ];
      
    public function mount() {
      
      // This does not work, erratic checkbox behavior, has to do with
      // the array elements being treated as string vs integer
      //$this->selections = Arr::pluck($this->data, 'id');  
      
      // This works.  Explicitly cast the array elements to strings
      $this->selections = array_map(function ($a) { return strval($a); }, Arr::pluck($this->data, 'id'));

    }

    public function render()
    {
        return view('hello-world');
    }

}
```

Livewire View:
```
<div>
<form action="#">
  @foreach($data as $datum)
    <input type="checkbox" wire:key="$datum['id']" value="{{$datum['id']}}" id="check-{{$datum['id']}}" wire:model="selections"/> {{$datum['text']}} -- {{$datum['id']}}<br>
  @endforeach
</form>
<hr>
@json($selections)<br>
@json($data)
</div>
```

Initially, in the `mount()` method of the component, I set the data of `$selections` by just plucking the id value out.  So, it was an array of integers.  When I tried checking/unchecking I had very erratic checkbox behavior (uncheck a box, uncheck a second one, and the first one comes back) and all the while, the bound data was not changing.  Eventually, I started noticing the `$selections` data started having a mix of strings and integers.  So, when I explicitly cast that initial array to all strings, everything started to work.

I tried using the `$casts` property to inform Livewire that these are integers like this:
```
    protected $casts = [
        'data.*.id' => 'integer',
    ];
```
... but to no effect.  I guessing either casts don't support nested data, or I just don't understand how they work.

I thought it might be connected to issue #793, but it's not data coming from PDO or any kind of database.  

Basically, I have a workaround so I can keep going.  But, I'm wondering if this is known, expected behavior, or if something is wrong.  If it is expected, it could probably use a note of some kind in the documentation (which I'd be glad to compose).  If not, I wondered if it would be worth a bug report.

Thanks everybody!

**Comment by jonneroelofs** [comment]
Played around with your example. That is a really weird problem.

Casts where deprecated in Livewire v2, so that is probably why you are not having any luck with those:
https://laravel-livewire.com/docs/2.x/upgrading#casters

Doesn't really answer you question, but this is how I would do it:
https://laravelplayground.com/#/snippets/f3116822-334d-4c32-8163-cb5135e711fc

**Reply by levi730** [reply]
Oops.  Totally missed that deprecation.  Thank you!

**Comment by jonneroelofs** [comment]
Looks like the problem is that a checkbox value is a string in the html spec. 

![image](https://user-images.githubusercontent.com/20800818/131120531-b67457b0-5073-404d-82a5-0e86aa2abc4e.png)
(https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#value)

**Reply by levi730** [reply]
Interesting.  Thank you.  I guess Livewire's method of comparison must be more nuanced that plain old PHP's.  I'll have to try to dig into how it does that comparison.  Thank you again!

**Comment by joshhanley** [comment]
@levi730 yeah @jonneroelofs has it right, the html spec returns strings.

So I personally find it best to convert my data to strings just as you did in mount, that way there isn't a mismatch in types.

Then you can always cast to an int when persisting, or I'm sure Laravel would probably handle a integer as a string anyway.

Hope this helps!
